File: src/readers/shapefile/mmap.ts
import { DataBaseFile } from './dbf';
import { MMapReader } from '../mmap';
import { ShapeFileReader } from './shp';
import { Transformer } from '../../proj4';
import { shapefileFromGzip } from '.';
import { exists, readFile } from 'fs/promises';

export * from './dbf';
export * from './shp';

/** A description of what relevant files exist and where */
export interface Definition {
  /** The path to the .shp file */
  shp: string;
  /** The path to the .dbf file. dbf is optional, but needed if you want attributes */
  dbf?: string;
  /**
   * The path to the .prj file. prj is optional, but needed if your file is in some
   * projection you don't want it in
   */
  prj?: string;
  /**
   * The path to the .cpg file. cpg is optional, but needed if your dbf is in some
   * weird (non utf8) encoding.
   */
  cpg?: string;
}

/**
 * # Build a Shapefile from an input path
 *
 * ## Description
 * Given a path to where all the shapefile relevant files exist, build a Shapefile
 *
 * Assumes the input is pointing to a shapefile or name without the extension.
 * The algorithm will find the rest of the paths if they exist. May also be a gzipped folder.
 *
 * ## Usage
 * ```ts
 * import { LambertConformalConic, EPSG_9974 } from 'gis-tools-ts';
 * import { shapefileFromPath } from 'gis-tools-ts/mmap';
 *
 * const reader = await shapefileFromPath('path/to/files', [LambertConformalConic], { EPSG_9974 });
 *
 * for await (const feature of reader) {
 *   console.log(feature);
 * }
 * ```
 * @param input - the path to the .shp file or name without the extension
 * @returns - a Shapefile
 */
export async function shapefileFromPath(input: string) {
  if (input.endsWith('.zip')) {
    const gzipData = await readFile(input);
    return shapefileFromGzip(gzipData.buffer);
  }
  const path = input.replace('.shp', '');
  const shp = `${path}.shp`;
  const dbf = `${path}.dbf`;
  const prj = `${path}.prj`;
  const cpg = `${path}.cpg`;
  if (!(await exists(shp))) throw new Error('Shapefile does not exist');
  const definition: Definition = {
    shp,
    dbf: (await exists(dbf)) ? dbf : undefined,
    prj: (await exists(prj)) ? prj : undefined,
    cpg: (await exists(cpg)) ? cpg : undefined,
  };
  return shapefileFromDefinition(definition);
}

/**
 * # Build a Shapefile from a Definition
 *
 * ## Description
 * Given a collection of files, build a Shapefile
 *
 * ## Usage
 * ```ts
 * import { LambertConformalConic, EPSG_9974 } from 'gis-tools-ts';
 * import { shapefileFromDefinition } from 'gis-tools-ts/mmap';
 * import type { Definition } from 'gis-tools-ts';
 *
 * const def: Definition = {
 *   shp: 'path/to/file.shp',
 *   dbf: 'path/to/file.dbf',
 *   prj: 'path/to/file.prj',
 *   cpg: 'path/to/file.cpg'
 * };
 *
 * const reader = await shapefileFromDefinition(def, [LambertConformalConic], { EPSG_9974 });
 *
 * for await (const feature of reader) {
 *   console.log(feature);
 * }
 * ```
 * @param def - a description of the data to parse
 * @returns - a Shapefile
 */
export async function shapefileFromDefinition(def: Definition): Promise<ShapeFileReader> {
  const { shp, dbf, prj, cpg } = def;
  const encoding = cpg !== undefined ? await readFile(cpg, { encoding: 'utf8' }) : 'utf8';
  const transform =
    prj !== undefined ? new Transformer(await readFile(prj, { encoding: 'utf8' })) : undefined;
  const dbfReader = dbf !== undefined ? new MMapReader(dbf) : undefined;
  const databaseFile = dbfReader !== undefined ? new DataBaseFile(dbfReader, encoding) : undefined;

  return new ShapeFileReader(new MMapReader(shp), databaseFile, transform);
}

File: rust/readers/shapefile/mmap.rs
use super::{DataBaseFile, Definition, ShapeFileReader};
use crate::{parsers::MMapReader, proj::Transformer};
use s2json::MValueCompatible;
use std::{
    collections::BTreeMap,
    fs::{File, exists},
    io::Read,
    path::Path,
    string::{String, ToString},
};

/// # Build a Shapefile from an input path
///
/// ## Description
/// Given a path to where all the shapefile relevant files exist, build a Shapefile
pub fn shapefile_from_path<I: AsRef<Path> + ToString, P: MValueCompatible>(
    input: I,
    epsg_codes: BTreeMap<String, String>,
) -> ShapeFileReader<MMapReader, P> {
    let path = input.to_string().replace(".shp", "");
    let shp = path.clone() + ".shp";
    let dbf_str = path.clone() + ".dbf";
    let prj_str = path.clone() + ".prj";
    let cpg_str = path.clone() + ".cpg";

    if exists(&shp).is_err() {
        panic!("Shapefile does not exist");
    }
    let dbf: Option<String> = if exists(&dbf_str).is_ok() { Some(dbf_str) } else { None };
    let prj: Option<String> = if exists(&prj_str).is_ok() { Some(prj_str) } else { None };
    let cpg: Option<String> = if exists(&cpg_str).is_ok() { Some(cpg_str) } else { None };
    let definition = Definition { shp, dbf, prj, cpg };

    shapefile_from_definition(definition, epsg_codes)
}

/// # Build a Shapefile from a Definition
///
/// ## Description
/// Given a collection of files, build a Shapefile
pub fn shapefile_from_definition<P: MValueCompatible>(
    def: Definition,
    epsg_codes: BTreeMap<String, String>,
) -> ShapeFileReader<MMapReader, P> {
    let Definition { shp, dbf, prj, cpg } = def;
    let mut database_file = None;
    let mut encoding = None;
    let mut transform = None;
    if let Some(cpg) = cpg {
        // read cpg file as string
        let mut file = File::open(cpg).unwrap();
        let mut input_str: String = String::new();
        let _ = file.read_to_string(&mut input_str);
        if !input_str.is_empty() {
            encoding = Some(input_str);
        }
    }
    // Handle projection
    if let Some(prj) = prj {
        let mut transformer = Transformer::new();
        for (code, value) in epsg_codes.iter() {
            transformer.insert_epsg_code(code.clone(), value.clone());
        }
        let pr_str = std::fs::read_to_string(prj).unwrap();
        transformer.set_source(pr_str);
        transform = Some(transformer);
    }
    // handle database data
    if let Some(dbf) = dbf {
        database_file = Some(DataBaseFile::new(MMapReader::from(dbf), encoding));
    }

    ShapeFileReader::new(MMapReader::from(shp), database_file, transform)
}
